// wechaty-puppet-xp/dist/cjs/examples/ripe-wechaty.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Wechaty - Conversational RPA SDK for Chatbot Makers.
 *  - https://github.com/wechaty/wechaty
 */
const wechaty_1 = require("wechaty");
const puppet_xp_js_1 = require("../src/puppet-xp.js");
const qrcode_terminal_1 = __importDefault(require("qrcode-terminal"));
const promises_1 = __importDefault(require("timers/promises"));
function onScan(qrcode, status) {
    if (qrcode) {
        const qrcodeImageUrl = [
            'https://wechaty.js.org/qrcode/',
            encodeURIComponent(qrcode),
        ].join('');
        console.info('StarterBot', 'onScan: %s(%s) - %s', status, qrcodeImageUrl);
        qrcode_terminal_1.default.generate(qrcode, { small: true }); // show qrcode on console
        console.info(`[${status}] ${qrcode}\nScan QR Code above to log in: `);
    }
    else {
        console.info(`[${status}]`);
    }
}
async function onLogin(user) {
    wechaty_1.log.info('StarterBot', '%s login', user);
    const roomList = await bot.Room.findAll();
    console.info(roomList.length);
}
function onLogout(user) {
    wechaty_1.log.info('StarterBot', '%s logout', user);
}
async function onMessage(msg) {
    wechaty_1.log.info('StarterBot', msg.toString());
    if (msg.text() === 'ding') {
        await msg.say('dong');
    }
    if (msg.type() === wechaty_1.types.Message.Image) {
        const img = await msg.toImage();
        const thumbFile = await img.thumbnail();
        wechaty_1.log.info('thumbFile', thumbFile.name);
        await thumbFile.toFile(`${process.cwd()}/cache/${thumbFile.name}`, true);
        await promises_1.default.setTimeout(1000);
        // console.info(img)
        const hdFile = await img.hd();
        wechaty_1.log.info('hdFile', hdFile.name);
        await hdFile.toFile(`${process.cwd()}/cache/${hdFile.name}`, true);
        setTimeout(msg.wechaty.wrapAsync(async function () {
            const imginfo = await msg.toFileBox();
            console.info(imginfo);
        }), 500);
    }
}
const puppet = new puppet_xp_js_1.PuppetXp();
const bot = wechaty_1.WechatyBuilder.build({
    name: 'ding-dong-bot',
    puppet,
});
bot.on('scan', onScan);
bot.on('login', onLogin);
bot.on('logout', onLogout);
bot.on('message', onMessage);
bot.start()
    .then(() => {
    return wechaty_1.log.info('StarterBot', 'Starter Bot Started.');
})
    .catch(console.error);
