// sidecar/dist/cjs/src/decorators/hook/hook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HOOK_TARGET_SYMBOL = exports.getMetadataHook = exports.Hook = void 0;
const config_js_1 = require("../../config.js");
const HOOK_TARGET_SYMBOL = Symbol('hookTarget');
exports.HOOK_TARGET_SYMBOL = HOOK_TARGET_SYMBOL;
function updateMetadataHook(target, propertyKey, functionTarget) {
    // Update the parameter names
    Reflect.defineMetadata(HOOK_TARGET_SYMBOL, functionTarget, target, propertyKey);
}
function getMetadataHook(target, propertyKey) {
    // Pull the array of parameter names
    const functionTarget = Reflect.getMetadata(HOOK_TARGET_SYMBOL, target, propertyKey);
    return functionTarget;
}
exports.getMetadataHook = getMetadataHook;
function Hook(functionTarget) {
    config_js_1.log.verbose('Sidecar', '@Hook(%s)', typeof functionTarget === 'object' ? JSON.stringify(functionTarget)
        : typeof functionTarget === 'number' ? functionTarget.toString(16)
            : functionTarget);
    return function hookMethodDecorator(target, propertyKey, descriptor) {
        config_js_1.log.verbose('Sidecar', '@Hook(%s) hookMethodDecorator(%s, %s, descriptor)', typeof functionTarget === 'object' ? JSON.stringify(functionTarget)
            : typeof functionTarget === 'number' ? functionTarget.toString(16)
                : functionTarget, target.constructor.name, propertyKey);
        updateMetadataHook(target, propertyKey, functionTarget);
        // Huan(202106) TODO: add a replaced function to show a error message when be called.
        return descriptor;
    };
}
exports.Hook = Hook;
