// sidecar/dist/cjs/src/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.SidecarBody = exports.Sidecar = exports.RetType = exports.Ret = exports.ParamType = exports.Hook = exports.exportTarget = exports.detach = exports.debug = exports.Call = exports.attach = exports.agentTarget = exports.addressTarget = void 0;
/**
 *   Sidecar - https://github.com/huan/sidecar
 *
 *   @copyright 2021 Huan LI (李卓桓) <https://github.com/huan>
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
const version_js_1 = require("./version.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_js_1.VERSION; } });
const ret_js_1 = require("./ret.js");
Object.defineProperty(exports, "Ret", { enumerable: true, get: function () { return ret_js_1.Ret; } });
const mod_js_1 = require("./sidecar-body/mod.js");
Object.defineProperty(exports, "SidecarBody", { enumerable: true, get: function () { return mod_js_1.SidecarBody; } });
Object.defineProperty(exports, "attach", { enumerable: true, get: function () { return mod_js_1.attach; } });
Object.defineProperty(exports, "detach", { enumerable: true, get: function () { return mod_js_1.detach; } });
const constants_js_1 = require("./sidecar-body/constants.js");
/**
 * Decorators
 */
const mod_js_2 = require("./decorators/call/mod.js");
Object.defineProperty(exports, "Call", { enumerable: true, get: function () { return mod_js_2.Call; } });
const mod_js_3 = require("./decorators/hook/mod.js");
Object.defineProperty(exports, "Hook", { enumerable: true, get: function () { return mod_js_3.Hook; } });
const mod_js_4 = require("./decorators/param-type/mod.js");
Object.defineProperty(exports, "ParamType", { enumerable: true, get: function () { return mod_js_4.ParamType; } });
const mod_js_5 = require("./decorators/ret-type/mod.js");
Object.defineProperty(exports, "RetType", { enumerable: true, get: function () { return mod_js_5.RetType; } });
const mod_js_6 = require("./decorators/sidecar/mod.js");
Object.defineProperty(exports, "Sidecar", { enumerable: true, get: function () { return mod_js_6.Sidecar; } });
/**
 * Target helpers
 */
const function_target_js_1 = require("./function-target.js");
Object.defineProperty(exports, "addressTarget", { enumerable: true, get: function () { return function_target_js_1.addressTarget; } });
Object.defineProperty(exports, "agentTarget", { enumerable: true, get: function () { return function_target_js_1.agentTarget; } });
Object.defineProperty(exports, "exportTarget", { enumerable: true, get: function () { return function_target_js_1.exportTarget; } });
const debug = {
    ATTACH_SYMBOL: constants_js_1.ATTACH_SYMBOL,
    DETACH_SYMBOL: constants_js_1.DETACH_SYMBOL,
    HOOK_EVENT_HANDLER: constants_js_1.HOOK_EVENT_HANDLER,
    INIT_SYMBOL: constants_js_1.INIT_SYMBOL,
    LOG_EVENT_HANDLER: constants_js_1.LOG_EVENT_HANDLER,
    SCRIPT_DESTROYED_HANDLER_SYMBOL: constants_js_1.SCRIPT_DESTROYED_HANDLER_SYMBOL,
    SCRIPT_MESSAGRE_HANDLER_SYMBOL: constants_js_1.SCRIPT_MESSAGRE_HANDLER_SYMBOL,
};
exports.debug = debug;
