// sidecar/dist/cjs/src/sidecar-body/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HOOK_EVENT_HANDLER = exports.LOG_EVENT_HANDLER = exports.SCRIPT_MESSAGRE_HANDLER_SYMBOL = exports.SCRIPT_DESTROYED_HANDLER_SYMBOL = exports.DETACH_SYMBOL = exports.ATTACH_SYMBOL = exports.INIT_SYMBOL = void 0;
const INIT_SYMBOL = Symbol('init');
exports.INIT_SYMBOL = INIT_SYMBOL;
const ATTACH_SYMBOL = Symbol('attatch');
exports.ATTACH_SYMBOL = ATTACH_SYMBOL;
const DETACH_SYMBOL = Symbol('detach');
exports.DETACH_SYMBOL = DETACH_SYMBOL;
const SCRIPT_DESTROYED_HANDLER_SYMBOL = Symbol('scriptDestroyedHandler');
exports.SCRIPT_DESTROYED_HANDLER_SYMBOL = SCRIPT_DESTROYED_HANDLER_SYMBOL;
const SCRIPT_MESSAGRE_HANDLER_SYMBOL = Symbol('scriptMessageHandler');
exports.SCRIPT_MESSAGRE_HANDLER_SYMBOL = SCRIPT_MESSAGRE_HANDLER_SYMBOL;
const LOG_EVENT_HANDLER = Symbol('logEventHandler');
exports.LOG_EVENT_HANDLER = LOG_EVENT_HANDLER;
const HOOK_EVENT_HANDLER = Symbol('hookEventHandler');
exports.HOOK_EVENT_HANDLER = HOOK_EVENT_HANDLER;
