// sidecar/dist/cjs/src/decorators/sidecar/build-sidecar-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildSidecarMetadata = void 0;
const config_js_1 = require("../../config.js");
const function_target_js_1 = require("../../function-target.js");
const metadata_call_js_1 = require("../call/metadata-call.js");
const hook_js_1 = require("../hook/hook.js");
const metadata_param_type_js_1 = require("../param-type/metadata-param-type.js");
const metadata_ret_type_js_1 = require("../ret-type/metadata-ret-type.js");
const target_js_1 = require("./target.js");
function buildSidecarMetadata(klass, options = {}) {
    config_js_1.log.verbose('Sidecar', 'buildSidecarMetadata(%s%s)', klass.name, options
        ? `, "${JSON.stringify(options).substr(0, 20)}..."`
        : '');
    const interceptorList = [];
    const nativeFunctionList = [];
    const propertyList = Object.getOwnPropertyNames(klass.prototype);
    for (const property of propertyList) {
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() building "%s.%s"...', klass.name, property);
        /**
         * 1. Get the metadata of `call` and `hook`
         */
        const callMetadata = (0, metadata_call_js_1.getMetadataCall)(klass.prototype, property);
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() callMetadata of %s: %s', property, JSON.stringify(callMetadata));
        const hookMetadata = (0, hook_js_1.getMetadataHook)(klass.prototype, property);
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() hookMetadata of %s: %s', property, JSON.stringify(hookMetadata));
        /**
         * 2. Make sure the target exists: either `call` or `hook`
         */
        const target = callMetadata || hookMetadata;
        if (!target) {
            config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() no callMetadata nor hookMetadata of %s: skip this loop', property);
            continue;
        }
        const targetObj = (0, function_target_js_1.normalizeFunctionTarget)(target);
        /**
         * 3. Get parameter and return types
         */
        const paramTypeMetadata = (0, metadata_param_type_js_1.getMetadataParamType)(klass.prototype, property);
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() paramTypeMetadata of %s: %s', property, JSON.stringify(paramTypeMetadata));
        const retTypeMetadata = (0, metadata_ret_type_js_1.getMetadataRetType)(klass.prototype, property);
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() retTypeMetadata of %s: %s', property, JSON.stringify(retTypeMetadata));
        /**
         * 4. Build the function descript and save it
         */
        const functionDescription = {
            [targetObj.type]: {
                name: property,
                paramTypeList: paramTypeMetadata,
                retType: retTypeMetadata,
                target: targetObj,
            },
        };
        config_js_1.log.silly('Sidecar', 'buildSidecarMetadata() functionDescription of %s: %s', property, JSON.stringify(functionDescription));
        if (callMetadata) {
            nativeFunctionList.push(functionDescription);
        }
        if (hookMetadata) {
            interceptorList.push(functionDescription);
        }
    }
    const { initAgentScript, sidecarTarget, } = options;
    const normalizedTarget = (0, target_js_1.normalizeSidecarTarget)(sidecarTarget);
    return {
        initAgentScript,
        interceptorList,
        nativeFunctionList,
        sidecarTarget: normalizedTarget,
    };
}
exports.buildSidecarMetadata = buildSidecarMetadata;
