// sidecar/dist/cjs/src/decorators/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sidecar = exports.RetType = exports.ParamType = exports.Name = exports.Hook = exports.Call = void 0;
const mod_js_1 = require("./call/mod.js");
Object.defineProperty(exports, "Call", { enumerable: true, get: function () { return mod_js_1.Call; } });
const mod_js_2 = require("./hook/mod.js");
Object.defineProperty(exports, "Hook", { enumerable: true, get: function () { return mod_js_2.Hook; } });
const mod_js_3 = require("./name/mod.js");
Object.defineProperty(exports, "Name", { enumerable: true, get: function () { return mod_js_3.Name; } });
const mod_js_4 = require("./param-type/mod.js");
Object.defineProperty(exports, "ParamType", { enumerable: true, get: function () { return mod_js_4.ParamType; } });
const mod_js_5 = require("./ret-type/mod.js");
Object.defineProperty(exports, "RetType", { enumerable: true, get: function () { return mod_js_5.RetType; } });
const mod_js_6 = require("./sidecar/mod.js");
Object.defineProperty(exports, "Sidecar", { enumerable: true, get: function () { return mod_js_6.Sidecar; } });
