// sidecar/dist/cjs/bin/sidecar-dump.js
#!/usr/bin/env node --no-warnings --loader ts-node/esm --experimental-vm-modules
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * https://github.com/huan/sidecar
 *
 * Author: Huan <zixia@zixia.net>
 * License: Apache-2.0
 *
 * CLI Apps in TypeScript with `cmd-ts` (Part 1)
 *  Using `cmd-ts` to easily build a type-safe TypeScript CLI app
 *
 *  https://gal.hagever.com/posts/type-safe-cli-apps-in-typescript-with-cmd-ts-part-1/
 */
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
// import {
//   REGISTER_INSTANCE,
// }                     from 'ts-node'
const version_js_1 = require("../src/version.js");
// import { log }      from '../src/config.js'
const mod_js_1 = require("../src/cli/mod.js");
/**
 * Check ts-node loaded or not
 *  See: https://github.com/TypeStrong/ts-node/blob/5643ad64cf39ee0dfa2a9323e8d1dd9f400e5884/src/index.ts#L54-L68
 *
 * Update:
 *  - Huan(202109): We enable ESM
 */
// if (!process[REGISTER_INSTANCE]) {
//   log.verbose('sidecar-dump', 'Loading `ts-node/register`...')
//   require('ts-node/register')
// }
const sidecarDump = (0, cmd_ts_1.subcommands)({
    name: 'sidecar-dump',
    description: 'Sidecar utility for dumping metadata/source for a sidecar class',
    version: version_js_1.VERSION,
    cmds: {
        metadata: mod_js_1.metadata,
        source: mod_js_1.source,
    },
});
(0, cmd_ts_1.run)((0, cmd_ts_1.binary)(sidecarDump), process.argv).catch(console.error);
