// sidecar/dist/cjs/src/wrappers/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapView = void 0;
const declare_js_args_js_1 = require("./declare-js-args.js");
const declare_native_args_js_1 = require("./declare-native-args.js");
const js_args_js_1 = require("./js-args.js");
const js_ret_js_1 = require("./js-ret.js");
const log_level_js_1 = require("./log-level.js");
const module_name_js_1 = require("./module-name.js");
const native_args_js_1 = require("./native-args.js");
const native_function_name_list_js_1 = require("./native-function-name-list.js");
const native_param_types_js_1 = require("./native-param-types.js");
const native_ret_type_js_1 = require("./native-ret-type.js");
const wrapView = (metadata) => ({
    ...metadata,
    declareJsArgs: declare_js_args_js_1.declareJsArgs,
    declareNativeArgs: declare_native_args_js_1.declareNativeArgs,
    jsArgs: js_args_js_1.jsArgs,
    jsRet: js_ret_js_1.jsRet,
    logLevel: log_level_js_1.logLevel,
    moduleName: module_name_js_1.moduleName,
    nativeArgs: native_args_js_1.nativeArgs,
    nativeFunctionNameList: native_function_name_list_js_1.nativeFunctionNameList,
    nativeParamTypes: native_param_types_js_1.nativeParamTypes,
    nativeRetType: native_ret_type_js_1.nativeRetType,
});
exports.wrapView = wrapView;
