// sidecar/dist/cjs/src/cli/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.metadata = void 0;
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
const fs_js_1 = require("cmd-ts/dist/cjs/batteries/fs.js");
const metadata_handler_js_1 = require("./metadata-handler.js");
async function handler(args) {
    const result = await (0, metadata_handler_js_1.metadataHandler)(args);
    // print the result
    console.log(result);
}
const metadata = (0, cmd_ts_1.command)({
    name: 'metadata',
    description: 'Dump sidecar metadata',
    args: {
        file: (0, cmd_ts_1.positional)({
            type: fs_js_1.File,
            displayName: 'classFile',
            description: 'The file contains the sidecar class',
        }),
        name: (0, cmd_ts_1.option)({
            description: 'The name of class that decorated by @Sidecar',
            long: 'name',
            short: 'n',
            type: (0, cmd_ts_1.optional)(cmd_ts_1.string),
        }),
    },
    handler,
});
exports.metadata = metadata;
