// sidecar/dist/cjs/src/cli/source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.source = void 0;
/* eslint-disable sort-keys */
const cmd_ts_1 = require("cmd-ts");
const fs_js_1 = require("cmd-ts/dist/cjs/batteries/fs.js");
const source_handler_js_1 = require("./source-handler.js");
async function handler(args) {
    const result = await (0, source_handler_js_1.sourceHandler)(args);
    console.log(result);
}
const source = (0, cmd_ts_1.command)({
    name: 'source',
    description: 'Dump sidecar agent source',
    args: {
        file: (0, cmd_ts_1.positional)({
            type: fs_js_1.File,
            displayName: 'classFile',
            description: 'The file contains the sidecar class',
        }),
        name: (0, cmd_ts_1.option)({
            description: 'The name of class that decorated by @Sidecar',
            long: 'name',
            short: 'n',
            type: (0, cmd_ts_1.optional)(cmd_ts_1.string),
        }),
    },
    handler,
});
exports.source = source;
